nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+
+### Validation and sanitization
+
+Before the data from a form is sent off to our server, we must first do two things
+
+- *Validation* ensures user input meets the specified criteria (e.g. required fields, correct format).
+- *Sanitization* cleans user input to prevent malicious data from being processed by removing or encoding potentially malicious characters.
+
+The library we'll be using is called `express-validator`. While it simplifies both of these processes for us, it's important to understand the underlying concepts of these two operations.
+
+#### Installation
+
+We start as usual by installing the correct package in the *root* folder of our project.
+
+```bash
+npm install express-validator
+```
+
+To get started using the package, we first need to *require* it in our project
+
+```javascript
+const { body, validationResult } = require("express-validator");
+```
+
+`express-validator` comes with many functions for just about every form operation you could think of, but for this lesson, we'll be sticking with `body()` and `validationResult()`.
+
+### Understanding the body() function
+
+The `body()` function allows you to specify which fields in the request body should be validated and sanitized, as well as how to handle it.
+
+```javascript
+[
+  body("birthdate", "Must be a valid date.")
+          .optional({ values: "falsy" })
+          .isISO8601() // Enforce a YYYY-MM-DD format.
+];
+```
+
+This example marks `birthdate` field as optional, but still enforces the ISO8601 date format on inputs. This is because `{ values: "falsy }"` means values that aren't `undefined`, `null`, `false`, or `0` will still be validated.
+
+### Chaining validations
+
+You can also chain multiple validation methods, with unique error messages if the checks fail.
+
+```javascript
+[
+    body("name")
+    .trim()
+    .isLength({ min: 1 })
+    .withMessage("Name can not be empty.")
+    .isAlpha()
+    .withMessage("Name must only contain alphabet letters."),  
+];
+```
+
+This ensures that `name` is not only present and trimmed, but also only contains alphabet letters.
+
+While this might work for outputs we know won't have special characters, like names or ages, we also have to consider situations that do allow those characters. For example, when writing their "About Me" description, what would happen if the client decides to inject JavaScript code instead?
+
+```ejs
+<div>
+  About Me: <%- description %>
+</div>
+
+// The client then inputs the following as their page's About Me:
+<script>alert("Hacked!");</script>
+```
+
+When unescaped, this would be rendered into HTML as:
+
+```html
+<div>
+  About Me: <script>alert("Hacked!");</script>!
+</div>
+```
+
+To prevent this cross-site scripting (XSS) attack, we can *escape* the output (you may also see this referred to as *encoding*). Escaped HTML replaces special characters, like `<`, with their respective HTML entities, in this case `&lt;`. In EJS, we can escape the output using `<%= %>`.
```

Comment:
nice inclusion of XSS

can we include a reference link for cross-site scripting attack, a MDN link or Wikipedia will do